File: frontend/src/api/userSettings.ts
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
import client from './client'
export interface NotificationToggle {
    app: boolean
    email: boolean
    feed: boolean
}

export interface UserSettings {
    emailEnabled: boolean
    theme: string
    darkMode: string
    notifications: Record<string, NotificationToggle>
    mailConfigured: boolean
    mailProviderName: string
    mailProviderUrl: string
}

export interface UserSettingsRequest {
    emailEnabled?: boolean
    theme?: string
    darkMode?: string
    feel?: string
    notifications?: Record<string, NotificationToggle>
}

export async function getSettings(): Promise<UserSettings> {
    const res = await client.get<UserSettings>('/settings')
    return res.data
}

export async function updateSettings(data: UserSettingsRequest): Promise<UserSettings> {
    const res = await client.put<UserSettings>('/settings', data)
    return res.data
}

File: src/main/java/dev/chojo/ember/feature/members/entity/UserSettings.java
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
package dev.chojo.ember.feature.members.entity;

import de.chojo.sadu.mapper.rowmapper.RowMapping;

/**
 * User-specific notification and communication settings.
 *
 * @param memberId     the station member identifier
 * @param emailEnabled whether email notifications are enabled for this member
 */
public record UserSettings(int memberId, boolean emailEnabled, String theme, String darkMode, String feel) {
    public static RowMapping<UserSettings> map() {
        return row -> new UserSettings(
                row.getInt("member_id"),
                row.getBoolean("email_enabled"),
                row.getString("theme"),
                row.getString("dark_mode"),
                row.getString("feel"));
    }
}
